File: crates/re_types/src/datatypes/class_description_map_elem_ext.rs
use super::{AnnotationInfo, ClassDescription, ClassDescriptionMapElem, Color};

impl From<(u16, &str)> for ClassDescriptionMapElem {
    fn from(value: (u16, &str)) -> Self {
        let class: ClassDescription = value.into();
        class.into()
    }
}

impl From<(u16, &str, Color)> for ClassDescriptionMapElem {
    fn from(value: (u16, &str, Color)) -> Self {
        let class: ClassDescription = value.into();
        class.into()
    }
}

impl From<AnnotationInfo> for ClassDescriptionMapElem {
    fn from(info: AnnotationInfo) -> Self {
        let class: ClassDescription = info.into();
        class.into()
    }
}

impl From<ClassDescription> for ClassDescriptionMapElem {
    fn from(class_description: ClassDescription) -> Self {
        Self {
            class_id: class_description.info.id.into(),
            class_description,
        }
    }
}

File: rerun_cpp/src/rerun/datatypes/class_description_map_elem_ext.cpp
#include <utility>
#include "class_description_map_elem.hpp"

// Uncomment for better auto-complete while editing the extension.
// #define EDIT_EXTENSION

namespace rerun {
    namespace datatypes {

#ifdef EDIT_EXTENSION
        struct ClassDescriptionMapElemExt {
            components::ClassId class_id;
            ClassDescription class_description;

#define ClassDescriptionMapElem ClassDescriptionMapElemExt

            // <CODEGEN_COPY_TO_HEADER>

            ClassDescriptionMapElem(ClassDescription _class_description)
                : class_id(_class_description.info.id),
                  class_description(std::move(_class_description)) {}

            // </CODEGEN_COPY_TO_HEADER>
        };
#endif

    } // namespace datatypes
} // namespace rerun
